server/src/main/java/org/elasticsearch/common/file/AbstractFileWatchingService.java

```diff
@@ -133,7 +133,11 @@ public final Path watchedFile() {
     @Override
     public final void clusterChanged(ClusterChangedEvent event) {
         ClusterState clusterState = event.state();
-        startIfMaster(clusterState);
+        if (clusterState.nodes().isLocalNodeElectedMaster()) {
+            startWatcher(clusterState);
+        } else if (event.previousState().nodes().isLocalNodeElectedMaster()) {
```

Comment:
Just to check, the `ClusterStateListener` is guaranteed to get every individual cluster change in order?